src/python/pants/core/goals/test.py

```diff
@@ -204,6 +223,135 @@ class TestFieldSet(FieldSet, metaclass=ABCMeta):
+                @rule
+                async def partition(
+                    request: SuperTesterRequest.PartitionRequest[SuperTesterFieldSet],
+                    subsystem: SuperTesterSubsystem,
+                ) -> Partitions[SuperTesterFieldSet]:
+                    if subsystem.skip:
+                        return Partitions()
+
+                    # One possible implementation
+                    return Partitions.partition_per_input(request.field_sets)
+
+        2. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestResult` instance. E.g.
+                @rule
+                async def super_test(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestResult:
+                    ...
+
+        3. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestDebugRequest` instance. E.g.
+                @rule
+                async def setup_super_test_debug(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestDebugRequest:
+                    ...
+
+        4. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestDebugAdapterRequest` instance. E.g.
+                @rule
+                async def setup_super_test_debug(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestDebugAdapterRequest:
+                    ...
+
+    NOTE: If your tester doesn't support running in debug mode / running with the debug adapter, you
+        can have the `@rule`s in 3 and 4 raise `NotImplementedError`s.
+
+    Lastly, register the rules which tell Pants about your plugin.
+    E.g.
+        def rules():
+            return [
+                *collect_rules(),
+                *SuperTesterRequest.registration_rules(),
+            ]
+    """
+
+    field_set_type: ClassVar[type[TestFieldSet]]
+
+    # Prevent this class from being detected by pytest as a test class.
+    __test__ = False
+
+    @distinct_union_type_per_subclass(in_scope_types=[EnvironmentName])
+    @dataclass(frozen=True)
+    @runtime_ignore_subscripts
+    class PartitionRequest(Generic[_FieldSetT]):
+        """Returns a unique `PartitionRequest` type per calling type.
+
+        This serves us 2 purposes:
+            1. `TestRequest.PartitionRequest` is the unique type used as a union base for plugin registration.
+            2. `<Plugin Defined Subclass>.PartitionRequest` is the unique type used as the union member.
+        """
+
+        field_sets: tuple[_FieldSetT, ...]
+
+    @distinct_union_type_per_subclass(in_scope_types=[EnvironmentName])
+    @dataclass(frozen=True)
+    @runtime_ignore_subscripts
+    class SubPartition(Generic[_FieldSetT]):
+        elements: Tuple[_FieldSetT, ...]
+        key: str
```

Comment:
Do you need this to be a string? Other partitioners use arbitrary classes (and I'm thinking of PRing allowing this type to have the `partition_description` as an attribute for a single-source-of-truth)

Reply:
Oh I see you have a `Partition` class which handles the description. I like that (at the cost of more boilerplate)
We should merge these implementations, not only for our own sanity, but for our plugin authors as well.

Can you file an issue, reference it here, and assign to milestone 2.15.x?

We can arm-wrestle on who gets the pleasure of implementing :smile: 

(but still same original question on `key` type)

Reply:
The way I've written things now, the `key` here gets threaded through to be the `extra_output_prefix` in the `TestResult` - having this as a `str` makes it easy. But maybe it'd be better to instead leave this as `Any` and have a dedicated field next to `description` for this 🤔

Reply:
> But maybe it'd be better to instead leave this as Any and have a dedicated field next to description for this

+1

Reply:
Maybe `Partition` should even be a superclass so plugins can subclass it and add on w/e extra fields might be needed 🤯

Reply:
I guess `extra_output_prefix` is technically a common thing for all `TestResult`s so it makes sense to expose for all plugins, hmm...

Reply:
#17094